src/transformers/utils/chat_template_utils.py

```diff
@@ -0,0 +1,305 @@
+# Copyright 2024 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import inspect
+import json
+import re
+from typing import Any, Union, get_args, get_origin, get_type_hints
+
+
+BASIC_TYPES = (int, float, str, bool, Any, type(None), ...)
+description_re = re.compile(r"^(.*?)[\n\s]*(Args:|Returns:|Raises:|\Z)", re.DOTALL)
+args_re = re.compile(r"\n\s*Args:\n\s*(.*?)[\n\s]*(Returns:|Raises:|\Z)", re.DOTALL)
+args_split_re = re.compile(r"(?:^|\n)\s*(\w+)\s*(?:\([\w\s\[\],.*]+\))?:\s*(.*?)\s*(?=\n\s*\w|\Z)", re.DOTALL)
+returns_re = re.compile(r"\n\s*Returns:\n\s*(.*?)[\n\s]*(Raises:|\Z)", re.DOTALL)
+
+
+def _get_json_schema_type(param_type):
+    if param_type == int:
+        return {"type": "integer"}
+    elif param_type == float:
+        return {"type": "number"}
+    elif param_type == str:
+        return {"type": "string"}
+    elif param_type == bool:
+        return {"type": "boolean"}
+    elif param_type == Any:
+        return {}
+    else:
+        return {"type": "object"}
+
+
+def _parse_type_hint(hint):
+    origin = get_origin(hint)
+    args = get_args(hint)
+
+    if origin is None:
+        return _get_json_schema_type(hint)
+
+    if origin is Union:
+        # If it's a union of basic types, we can express that as a simple list in the schema
+        if all(t in BASIC_TYPES for t in args):
+            return_dict = {"type": [_get_json_schema_type(t)["type"] for t in args if t not in (type(None), ...)]}
+            if len(return_dict["type"]) == 1:
+                return_dict["type"] = return_dict["type"][0]
+        else:
+            # A union of more complex types requires us to recurse into each subtype
+            return_dict = {
+                "anyOf": [_parse_type_hint(t) for t in args if t not in (type(None), ...)],
```

Comment:
possibly because I'm recursion-averse :smile:  but I prefer stack-based traversals, I don't know if you tried something like that with a stack that'd be initd as 
```python
stack = [(hint, None)] 
# which you can pop current_hint and parent, then in while loop
current_hint, parent = stack.pop()
origin = get_origin(current_hint)
args = get_args(current_hint)


```
So if origin is `None`, don't add to the stack, else go through the other types